### PR TITLE
Add Pandoc version requirement in Lua Filters

### PR DESCRIPTION
**Pull request by atusy**
This PR requires Lua filters to be run on Pandoc >= 2.1.
The filters are skipped if requirements are unmet.

<details><summary>outdated initial comment</summary>
There are currently 3 filters available in **rmarkdown**

1. `latex-div.lua`
2. `number-sections.lua`
3. `pagebreak.lua`

As far as I understand, 1 and 2 requires Pandoc 2.0.0 and 3 requires Pandoc 2.0.6.
However, there are no way to check the version in Lua filter Pandoc < 2.1.0.
I decided to comment minimal requested version in the filters.
I also tweaked `pagebreak.lua` so that it does nothing when `pandoc.utils` is missing.
</details>

### CODE REVIEW

**Comment by atusy** [comment]
Thanks @cderv for the comment https://github.com/rstudio/rmarkdown/issues/2082#issuecomment-811909826

> Pandoc 2.1 for using Lua filters as I don't think this is very sensible because should be a few users (if none) 2.0 and 2.0.6.

I'm facing some trouble with supporting Pandoc < 2.1, so I will follow your suggestion.

**Comment by cderv** [comment]
Yes I'll review it. @atusy is the PR ready ? I was under the impression you were still working on it. 
Don't hesitate to open as draft and change to ready when you are ready for a review.

**Comment by atusy** [comment]
@yihui @cderv Thanks. I'm still working on this. I turned the PR to draft, and will be back soon.

**Comment by atusy** [comment]
Okay, I think I'm ready now.

Note thtat R CMD Check on macOS has not finished, and is keep saying "Starting worflow run..."
https://github.com/rstudio/rmarkdown/runs/2246561590

**Review by yihui — APPROVED** [review]
Okay, the alternative implementation looks even better to me.

**Comment by cderv** [comment]
I'll look at it as soon as possible and merge it. Thanks !

File: inst/rmarkdown/lua/latex-div.lua
Comment (cderv) [inline inst/rmarkdown/lua/latex-div.lua:8]: After some thoughts I would use `io.stderr:write` for these. We may also choose a layout that make it unique and recognizable as a warning / log string

Something like 
```
[WARNING](from latex-div.lua) requires Pandoc > 2.1. Lua filter has been ignored and some features disabled
```
What do you think ? 
```suggestion
    io.stderr:write("[WARNING] (latex-div.lua) requires at least Pandoc 2.1")
```

**Review by cderv — CHANGES_REQUESTED** [review]
Thanks ! I wanted to check for how PANDOC_VERSION works. 
This would be clean but it raises an error which we don't want.
```
PANDOC_VERSION:must_be_at_least '2.1'
```
Ignoring is ok I would just use `stderr` instead of `stdout` after some thoughts.
What do you think ? 

Also, we have a vignette about Lua Filter that we may need to update to add about the version requirement.

File: .github/workflows/R-CMD-check.yaml
Comment (atusy) [reply .github/workflows/R-CMD-check.yaml:34]: because 2.0 provides `pandoc-2.0-3-amd64.deb`, which does not match `pandoc-{{version}}-1-amd64.deb`.
However, I just noticed 2.0.0.1 is the minimal version that matches the above format.
Maybe 2.0.0.1 is the choice.

https://github.com/jgm/pandoc/releases?after=2.1

File: inst/rmarkdown/lua/latex-div.lua
Comment (atusy) [reply inst/rmarkdown/lua/latex-div.lua:8]: I like it

File: .github/workflows/R-CMD-check.yaml
Comment (cderv) [reply .github/workflows/R-CMD-check.yaml]: Oh I see it is a `r-lib/actions/setup-pandoc` limitation. 

It seems Pandoc 2.0.0 has not the same format. Earlier versions seems to have one. We could also look at other os. 

What is the aim of adding this version ? Testing that the tests are skipped ? 
I was under the impression you suggested to add a CI for Pandoc < 2.0 initially

Comment (atusy) [reply .github/workflows/R-CMD-check.yaml]: I'm currently testing 2.0 and 2.0.0.1 with GitHub Actions.
Let's employ the minimal install-able version.
61fb7e9

File: .github/workflows/R-CMD-check.yaml
Comment (cderv) [inline .github/workflows/R-CMD-check.yaml:34]: But this won't work as you explained ? File is not in the good format for the action to download. 
https://github.com/jgm/pandoc/releases/tag/2.0

Also I don't want us to test all Pandoc version. We are running CI regularly and it becomes a lot of running times (and resources). 

Why is our aim with that ? 

We could also add a Weekly or Monthly run dedicated to Pandoc version on only one version of R. This would reduce the load for each commit or PR.

File: inst/rmarkdown/lua/latex-div.lua
Comment (cderv) [reply inst/rmarkdown/lua/latex-div.lua]: To add to that, I am wondering if it would be useful to add more info on the message. like 

````
io.stderr:write("[WARNING] (latex-div.lua) requires at least Pandoc 2.1. Lua Filter has been ignored and some features hence disabled."
```

Too much ?

File: .github/workflows/R-CMD-check.yaml
Comment (atusy) [reply .github/workflows/R-CMD-check.yaml:34]: I feel it is reasonable to test minimal 2.x version.
At the same time, I agree it is better to avoid testing too many versions for every commits and PRs.
The addition of 2.0 is the temporary trial, and I will remove it.
I am okay to remove 2.0.0.1 as well.

File: .github/workflows/R-CMD-check.yaml
Comment (atusy) [reply .github/workflows/R-CMD-check.yaml]: Sorry, I missed your latest comment.
About Pandoc < 2.0, I've tested it but raising error which is not related to lua filter.
I think we should fix it and add to test in the different PR.

> Error: Error: pandoc document conversion failed with error 83
> https://github.com/atusy/rmarkdown/pull/2/checks?check_run_id=2219792064

File: inst/rmarkdown/lua/latex-div.lua
Comment (atusy) [reply inst/rmarkdown/lua/latex-div.lua]: I agree previous warning message is not kind enough because it does not explain what happens.
I think we can simply say "skipping" the filer, which implies disabling some features.

```
io.stderr:write("[WARNING] Skipping (latex-div.lua) which requires at least Pandoc 2.1."
```

File: .github/workflows/R-CMD-check.yaml
Comment (cderv) [reply .github/workflows/R-CMD-check.yaml]: yeah we can add one. I'll open a new issue to rediscuss which version we want to test. 

I did 2.7.3 mainly because it was the default in the RStudio IDE but it has now upgraded.

File: inst/rmarkdown/lua/latex-div.lua
Comment (cderv) [reply inst/rmarkdown/lua/latex-div.lua]: Like the requirement first which is the cause, then the effect it has. 
Maybe 
```
io.stderr:write("[WARNING] (latex-div.lua) requires at least Pandoc 2.1. Lua Filter skipped")
```

File: .github/workflows/R-CMD-check.yaml
Comment (atusy) [reply .github/workflows/R-CMD-check.yaml]: great. thanks for your neat comments.

File: inst/rmarkdown/lua/latex-div.lua
Comment (atusy) [reply inst/rmarkdown/lua/latex-div.lua]: got it, and done it by 2105f6d.

**Comment by cderv** [comment]
Thanks! 

As a sum up: 

* `pagebreak.lua` requires Pandoc >= 2.1 because it uses `pandoc.utils`
* Other lua filter have no specific requirement **but** we made a requirement for Pandoc >= 2.1  because we use `PANDOC_VERSION` in every filter ?

Is that correct ? 

I'll merge the change after some last tweaks and we'll see if we have feedback from people stuck between 2.0 and 2.1. 

After looking it is between 2017-10-29 and 2018-01-07 - so it is 2 months only. It should be fine. Nobody reported issue as Pandoc should have error with the pagebreak filter.

**Review by cderv — APPROVED** [review]

**Comment by atusy** [comment]
Thank you, too, for the nice reviews!
And yes, your sum up is correct!!

**Comment by jeroen** [comment]
I think this PR is the cause of vignette errors: https://github.com/rstudio/rmarkdown/issues/2099

**Pull request merged**
